File: Data Structures & Algorithms/maximum-product-subarray/submission-5.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res = INT_MIN;

        int currMax = 1;
        int currMin = 1;
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] < 0) {
                swap(currMax, currMin);
            }
            int nextMax = currMax * nums[i];
            int nextMin = currMin * nums[i];

            currMax = max(nextMax, nums[i]);
            currMin = min(nextMin, nums[i]);

            res = max(currMax, res);
        }

        return res;
    }
};

File: Data Structures & Algorithms/jump-game/submission-5.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        maxIndex = 0
        i = 0

        while i <= len(nums) and i <= maxIndex:
            maxIndex = max(i + nums[i], maxIndex)
            if maxIndex >= len(nums) - 1:
                return True

            i += 1
        
        return False